### PR TITLE
enhancement(sinks): Add acknowledgement configuration to sinks

### LINKED ISSUE #10983: Move end-to-end acknowledgement setting to sinks
Currently, end-to-end acknowledgements are opt-in at the source-level via the `acknowledgements.enabled` setting. This made sense initially since sources are the ones that acknowledging back to clients, but after discussing  #10273 we've decided that this should be a sink-level option instead. This makes it easier for users to achieve durability which is sink dependent not source dependent. For example, a user only cares that data is not lost going to their system of record, they do not care which source it came from, just that we guarantee delivery to the destination.

- [x] Move the `acknowledments` setting to the sinks instead.
- [x] Log a warning if any connected upstream sources do not support acknowledgements.
- [x] Deprecate the source-level `acknowledgements` setting.
- [x] Error if sinks that don't support acknowledgements have acknowledgments configured
- [x] Document e2e acknowledgements

### PR DESCRIPTION
**Pull request by bruceg**
Closes #10983 

Note that the documentation for this new feature is currently disabled AFAICT. There is a bit of a chicken-and-egg situation there with #7436 which actually adds the acknowledgements documentation for all the relevant sources and sinks. As such I decided to leave the docs deactivated to be later enabled by that PR.

### CODE REVIEW

**Review by jszwedko — COMMENTED** [review]
A question that just occurred to me: will this work enable sources connected to multiple sinks to ack as soon as associated sinks that have `acknowledgements = true` but not wait for sinks with `acknowledgements = false`?

The scenario I'm thinking about is a user hooking up source A to sinks B and C, but only needing B to ack, but allowing C to be best effort. This scenario is hinted at in https://github.com/vectordotdev/vector/issues/10983 but not explicitly outlined.

**Comment by bruceg** [comment]
> will this work enable sources connected to multiple sinks to ack as soon as associated sinks that have `acknowledgements = true` but not wait for sinks with `acknowledgements = false`?

No, this does not change that behavior. To make that work, I will need to change the internals of the finalizer types to send notification at some point other than at drop, possibly using channels instead of just atomics. That would be a significantly larger effort.

**Review by lukesteensen — APPROVED** [review]
Looks good to me!

> To make that work, I will need to change the internals of the finalizer types to send notification at some point other than at drop, possibly using channels instead of just atomics. That would be a significantly larger effort.

I only have a passing familiarity with the finalizer internals, but I wonder if you could do something along the lines of `Arc::downgrade` in non-acking sinks to let the remaining acking sinks control when `Drop` happens.

**Comment by bruceg** [comment]
> I only have a passing familiarity with the finalizer internals, but I wonder if you could do something along the lines of `Arc::downgrade` in non-acking sinks to let the remaining acking sinks control when `Drop` happens.

🤔 That might work. I'd have to actually put it into the code to be sure. That seems like it would be more efficient than a channel, but could be more invasive in terms of code changes.

File: src/config/mod.rs [hunk 1]
```diff
@@ -111,6 +112,64 @@ pub struct Config {
...
+            let component = &input.component;
+            if let Some(source) = self.sources.get_mut(component) {
+                if source.inner.can_acknowledge() {
+                    source.sink_acknowledgements = true;
+                } else {
+                    warn!(
+                        message = "Source has acknowledgements enabled by sink, but acknowledgements are not supported.",
+                        source = component.id(),
+                        sink = sink.id(),
+                    );
```
Comment (tobz) [inline src/config/mod.rs:159]: 👍🏻

**Review by tobz — APPROVED** [review]
LGTM, with one non-blocking question/nit.

File: src/config/mod.rs [hunk 2]
```diff
@@ -111,6 +112,64 @@ pub struct Config {
...
+
+    fn propagate_acks_rec(&mut self, sink_inputs: Vec<(ComponentKey, OutputId)>) {
+        for (sink, input) in sink_inputs {
+            let component = &input.component;
+            if let Some(source) = self.sources.get_mut(component) {
+                if source.inner.can_acknowledge() {
+                    source.sink_acknowledgements = true;
+                } else {
+                    warn!(
+                        message = "Source has acknowledgements enabled by sink, but acknowledgements are not supported.",
```
Comment (jszwedko) [inline src/config/mod.rs:156]:
```suggestion
                        message = "Source has acknowledgements enabled by a sink, but acknowledgements are not supported by this source. Data loss could occur.",
```

File: src/config/mod.rs [hunk 3]
```diff
@@ -164,13 +223,20 @@ pub struct SourceOuter {
     pub proxy: ProxyConfig,
     #[serde(flatten)]
     pub(super) inner: Box<dyn SourceConfig>,
+    #[serde(
+        default,
+        skip_deserializing,
+        skip_serializing_if = "vector_core::serde::skip_serializing_if_default"
```
Comment (jszwedko) [inline src/config/mod.rs:229]: We should always skip serializing, no?

File: src/sources/vector/mod.rs [hunk 4]
```diff
@@ -88,6 +88,10 @@ impl SourceConfig for VectorConfig {
             VectorConfig::V2(v2) => v2.config.resources(),
         }
     }
+
+    fn can_acknowledge(&self) -> bool {
```
Comment (jszwedko) [inline src/sources/vector/mod.rs:92]: Technically only `v2` can acknowledge. I think it's ok to leave as-is though as `v1` is deprecated and will be going away in a release or two.

File: website/cue/reference/components/sinks.cue [hunk 5]
```diff
@@ -6,6 +6,25 @@ components: sinks: [Name=string]: {
 	features: _
 
 	configuration: {
+		if features.acknowledgements != _|_ {
+			if features.acknowledgements.enabled {
+				acknowledgements: {
+					common:      true
+					description: "Controls how acknowledgements are handled by this sink. When enabled, all connected sources that support end-to-end acknowledgements will wait for the destination of this sink to acknowledge receipt of events before providing acknowledgement to the sending source. These settings override the global `acknowledgement` settings."
```
Comment (jszwedko) [inline website/cue/reference/components/sinks.cue:13]: Nit I'd use `"""` to wordwrap this.

File: website/cue/reference/components/sinks.cue [hunk 6]
```diff
@@ -6,6 +6,25 @@ components: sinks: [Name=string]: {
...
+					required:    false
+					type: object: options: {
+						enabled: {
+							common:      true
+							description: "Controls if all connected sources will wait for this sink to deliver the events before acknowledging receipt."
+							warnings: ["Disabling this option may lead to loss of data, as destination sinks may reject events after the source acknowledges their successful receipt."]
```
Comment (jszwedko) [inline website/cue/reference/components/sinks.cue:19]: I might leave this warning out until we make it the default. Or change the wording a bit to be something like "We recommend enabling this...".

File: src/config/mod.rs [hunk 7]
```diff
@@ -188,6 +254,10 @@ pub trait SourceConfig: core::fmt::Debug + Send + Sync {
     fn resources(&self) -> Vec<Resource> {
         Vec::new()
     }
+
+    fn can_acknowledge(&self) -> bool {
+        false
+    }
```
Comment (jszwedko) [inline src/config/mod.rs:260]: I'm inclined to leave this default implementation out to force new component implementations to consider whether it should be true or false. I could see us forgetting to add an implementation to new sources that do support it.

**Review by jszwedko — COMMENTED** [review]
Just a few comments, but generally looks good! I agree with Toby's comments too.

File: src/config/mod.rs [hunk 8]
```diff
@@ -226,7 +298,19 @@ impl SourceContext {
             shutdown: ShutdownSignal::noop(),
             out,
             proxy: Default::default(),
+            acknowledgements: false,
+        }
+    }
+
+    pub fn do_acknowledgements(&self, config: &AcknowledgementsConfig) -> bool {
+        if config.enabled() {
+            warn!("The `acknowledgements` configuration on sources is deprecated and will be removed in a future version.");
```
[the remark this answers is not in the document]
Comment (bruceg) [reply src/config/mod.rs]: Wording changed in f4c88cc0e. I don't think I can actually put a link in here, though, as the cue docs won't be rendering the text properly until we get the other docs changes merged in.

File: src/config/mod.rs (hunk 2 above)
Comment (bruceg) [reply src/config/mod.rs]: Done in 30b732d76

File: src/config/mod.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (bruceg) [reply src/config/mod.rs]: Done in b67e7ea8c

File: src/sources/vector/mod.rs (hunk 4 above)
Comment (bruceg) [reply src/sources/vector/mod.rs:92]: Yeah, I'm not sure how I could encode that easily, and didn't think it was a big deal given the deprecation.

File: src/config/mod.rs (hunk 3 above)
Comment (jszwedko) [reply src/config/mod.rs:229]: I guess maybe this shouldn't even be annotated as `serde`? It should never be read from the configuration, but rather set by sinks?

Comment (bruceg) [reply src/config/mod.rs:229]: The trick is that the sink acknowledgements are propagated at the config level, while the `SourceContext` is only created on-the-fly as sources are built. I _think_ the only place I can store this data is at the `SourceOuter` level. I have rewritten this to `#[serde(skip)]` which means serde just ignores it and resets it to default.

**Review by jszwedko — APPROVED** [review]

**Review by lukesteensen — APPROVED** [review]

**Pull request merged**